drivers/hwinfo/hwinfo_stm32.c

```diff
@@ -29,3 +29,72 @@ ssize_t z_impl_hwinfo_get_device_id(u8_t *buffer, size_t length)
 
 	return length;
 }
+
+int z_impl_hwinfo_get_reset_cause(u32_t *cause)
+{
+	u32_t flags = 0;
+
+#if defined(RCC_FLAG_SFTRST)
+	if (LL_RCC_IsActiveFlag_SFTRST()) {
+		flags |= RESET_SOFTWARE;
+	}
+#endif
+#if defined(RCC_FLAG_PINRST)
+	if (LL_RCC_IsActiveFlag_PINRST()) {
+		flags |= RESET_PIN;
+	}
+#endif
+#if defined(RCC_FLAG_IWDGRST)
+	if (LL_RCC_IsActiveFlag_IWDGRST()) {
+		flags |= RESET_INDEPENDENT_WATCHDOG;
+	}
+#endif
+#if defined(RCC_FLAG_WWDGRST)
+	if (LL_RCC_IsActiveFlag_WWDGRST()) {
+		flags |= RESET_WINDOW_WATCHDOG;
+	}
+#endif
+#if defined(RCC_FLAG_LPWRRSTF)
+	if (LL_RCC_IsActiveFlag_LPWRRST()) {
+		flags |= RESET_LOW_POWER_ERROR;
+	}
+#endif
+#if defined(RCC_FLAG_FWRST)
+	if (LL_RCC_IsActiveFlag_FWRST()) {
+		flags |= RESET_FIREWALL;
+	}
+#endif
+#if defined(RCC_FLAG_OBLRST)
+	if (LL_RCC_IsActiveFlag_OBLRST()) {
+		flags |= RESET_OPTION_BYTE_LOADER;
+	}
+#endif
+#if defined(RCC_FLAG_BORRST)
+	if (LL_RCC_IsActiveFlag_BORRST()) {
+		flags |= RESET_BROWNOUT;
+	}
+#endif
+#if defined(RCC_FLAG_PWRRST)
+	if (LL_RCC_IsActiveFlag_PWRRST()) {
+		flags |= RESET_POWER;
+	}
+#endif
```

Comment:
Correct. There will be some platform specific behaviour which the user will have to account for in her code. 

I considered setting both flags for this platform, but then you could a false trigger of brownout code path in the user's code when it was a POR. It can go both ways because the SoC flag itself is ambiguous. I decided to keep it simple.